### PR TITLE
Add auth and version support to socks proxy

### LINKED ISSUE #785: [Bug] Specify Authentication for SOCKS 5 Proxy
G'day love the app

I was looking to set it up to use a socks proxy that required auth and found that there is not config option to do so. I also tried setting the url to username:password@proxyserver with no luck

Is this a bug or is this just not a feature of the app yet?

Cheers

### LINKED ISSUE #857: [Bug] Extension still listed in "all"
## Device information
- Suwayomi-Server version: Server: v0.7.0 + WebUI: r983
- Server Operating System: Win 10 64bit
- Server Desktop Environment: Any
- Server JVM version: bundled with win32
- Client Operating System: win 10 64bit
- Client Web Browser: Firefox

## Steps to reproduce
1. Install an extension
2. Uninstall extension
3. Extension is still listed under "all" with an "install" button (but cant be installed, since the apk was removed with the uninstallation of course)

### Expected behavior
Is not listed anymore in "all" or anywhere else.

### Actual behavior
It is still listed in "all". how to completely remove an extension from Suwamayomi's memory?

### PR DESCRIPTION
**Pull request by AriaMoradi**
closes #785

I didn't test the code against a real proxy but it should be fine according to the docs.

Java 8: https://docs.oracle.com/javase/8/docs/technotes/guides/net/properties.html#socks
Java 20 (much more readable): https://docs.oracle.com/en/java/javase/20/core/java-networking.html#GUID-E0E7F3CF-FF30-4324-B2F8-AA88EB067994

### CODE REVIEW

File: server/src/main/kotlin/suwayomi/tachidesk/server/ServerSetup.kt [hunk 1]
```diff
@@ -236,14 +236,25 @@ fun applicationSetup() {
...
+            serverConfig.socksProxyPassword,
+        ) { proxyEnabled, proxyHost, proxyPort, proxyUsername, proxyPassword ->
+            data class DataClassForDestruction(
+                val proxyEnabled: Boolean,
+                val proxyHost: String,
+                val proxyPort: String,
+                val proxyUsername: String,
+                val proxyPassword: String,
+            )
+            DataClassForDestruction(proxyEnabled, proxyHost, proxyPort, proxyUsername, proxyPassword)
```
Comment (AriaMoradi) [inline server/src/main/kotlin/suwayomi/tachidesk/server/ServerSetup.kt]: not sure if this is the best way to convert multiple values to one, kotlin doesn't have a Quintuple type.

**Comment by Robonau** [comment]
i think u linked the wrong issue

File: server/src/main/kotlin/suwayomi/tachidesk/server/ServerSetup.kt [hunk 2]
```diff
@@ -236,14 +236,27 @@ fun applicationSetup() {
...
+                "Socks Proxy changed - enabled=$proxyEnabled address=$proxyHost:$proxyPort , username=$proxyUsername, password=[REDACTED]",
+            )
             if (proxyEnabled) {
                 System.getProperties()["socksProxyHost"] = proxyHost
                 System.getProperties()["socksProxyPort"] = proxyPort
+                System.getProperties()["java.net.socks.username"] = proxyUsername
+                System.getProperties()["java.net.socks.password"] = proxyPassword
             } else {
                 System.getProperties()["socksProxyHost"] = ""
                 System.getProperties()["socksProxyPort"] = ""
```
Comment (Syer10) [inline server/src/main/kotlin/suwayomi/tachidesk/server/ServerSetup.kt]:
```suggestion
            if (proxyEnabled) {
                System.setProperty("socksProxyHost", proxyHost)
                System.setProperty("socksProxyPort", proxyPort)
                if (proxyUsername.isNotBlank()) {
                    System.setProperty("java.net.socks.username", proxyUsername)
                } else {
                    System.clearProperty("java.net.socks.username")
                }
                if (proxyPassword.isNotBlank()) {
                    System.setProperty("java.net.socks.password", proxyPassword)
                } else {
                    System.clearProperty("java.net.socks.password")
                }
            } else {
                System.clearProperty("socksProxyHost")
                System.clearProperty("socksProxyPort")
```

File: server/src/main/kotlin/suwayomi/tachidesk/server/ServerSetup.kt [hunk 3]
```diff
@@ -236,14 +236,27 @@ fun applicationSetup() {
...
+        ) { proxyEnabled, proxyHost, proxyPort, proxyUsername, proxyPassword ->
+            data class DataClassForDestruction(
+                val proxyEnabled: Boolean,
+                val proxyHost: String,
+                val proxyPort: String,
+                val proxyUsername: String,
+                val proxyPassword: String,
+            )
+            DataClassForDestruction(proxyEnabled, proxyHost, proxyPort, proxyUsername, proxyPassword)
         },
```
Comment (Syer10) [inline server/src/main/kotlin/suwayomi/tachidesk/server/ServerSetup.kt]:
```suggestion
            serverConfig.socksProxyUsername,
            serverConfig.socksProxyPassword,
        ) { proxyEnabled, proxyHost, proxyPort, proxyUsername, proxyPassword ->
            data class ProxySettings(
                val proxyEnabled: Boolean,
                val proxyHost: String,
                val proxyPort: String,
                val proxyUsername: String,
                val proxyPassword: String,
            )
            ProxySettings(proxyEnabled, proxyHost, proxyPort, proxyUsername, proxyPassword)
        }.distinctUntilChanged(),
```

**Review by Syer10 — CHANGES_REQUESTED** [review]
Looks good, just some general cleanup

**Comment by AriaMoradi** [comment]
Might as well add support for socks 4 by providing java's `socksProxyVersion`

File: server/src/main/kotlin/suwayomi/tachidesk/server/ServerSetup.kt (hunk 3 above)
Comment (AriaMoradi) [reply server/src/main/kotlin/suwayomi/tachidesk/server/ServerSetup.kt]: API says it doesn't need `distinctUntilChanged()`
image

Comment (Syer10) [reply server/src/main/kotlin/suwayomi/tachidesk/server/ServerSetup.kt]: Its just to be safe.

**Review by Syer10 — APPROVED** [review]

**Pull request merged**